In [1]:
from opr_structs import *
from opr_logic import *
from pprint import pprint
import copy

In [2]:
import logging
logger = logging.getLogger()
logger.setLevel(logging.DEBUG)
logging.debug("test")
logger.setLevel(logging.WARNING)

DEBUG:root:test


In [3]:
from contextlib import contextmanager

@contextmanager
def loggingcm(level=logging.DEBUG):
    # Code to acquire resource, e.g.:
    logger.setLevel(level)
    try:
        yield
    finally:
        # Code to release resource, e.g.:
        logger.setLevel(logging.WARNING)

with loggingcm():
    logger.debug("debug1")
logger.debug("debug2")

DEBUG:root:debug1


In [4]:
def build_dwarven_army(playername):
    units = []
    
    # equipment
    hand_weapon = Weapon(1)
    rifle = Weapon(1, {"Range": 18, "AP": 1})
    champion_hand_weapon = Weapon(3, {"AP": 1})
    
    # Dwarf Champion
    units.append(Unit(
        [Model(3, 4, [champion_hand_weapon], rules={"Fearless":0, "Hero":0, "Slow":0, "Tough":3, "Fear":3})], 1, name="Hero"))
    
    dwarf_warrior = Model(4, 5, [hand_weapon], rules={"Shield Wall":0, "Slow":0})
    units.append(Unit(models_for(dwarf_warrior, 10), 5, name="Warriors"))
    
    heavy_hand_weapon = Weapon(1, {"AP": 1})
    veteran = Model(3, 4, [heavy_hand_weapon], rules={"Fearless":0, "Slow":0})
    units.append(Unit(models_for(veteran, 10), 5, name="Veterans"))
    
    great_weapon = Weapon(2, {"AP": 2})
    hammerer = Model(3, 4, [great_weapon], rules={"Slow":0})
    units.append(Unit(models_for(hammerer, 5), 5, name="Hammerers"))
    
    rifle_dwarf = Model(4, 5, [rifle, hand_weapon], rules={"Slow":0})
    units.append(Unit(models_for(rifle_dwarf, 5), 5, name="Marksmen"))
    
    fire_rifle = Weapon(2, {"Range": 18, "Rending": 0})
    drake_marksman = Model(3, 3, [fire_rifle, hand_weapon], rules={"Slow":0})
    units.append(Unit(models_for(drake_marksman, 5), 5, name="Drakes"))
    
    crew = Weapon(3)
    bolt_thrower = Weapon(1, {"Range": 36, "AP": 3, "Deadly":6})
    units.append(Unit([Model(4, 5, [crew, bolt_thrower], rules={"Immobile":0, "Entrenched":0, "Tough":3})], 1, name="Arty"))
    
    return units
build_dwarven_army("dwarves")

[Unit(Hero (1/1)(0.0,0.0),
 Unit(Warriors (10/10)(0.0,0.0),
 Unit(Veterans (10/10)(0.0,0.0),
 Unit(Hammerers (5/5)(0.0,0.0),
 Unit(Marksmen (5/5)(0.0,0.0),
 Unit(Drakes (5/5)(0.0,0.0),
 Unit(Arty (1/1)(0.0,0.0)]

In [5]:
hand_weapon = Weapon(1)

champion_hand_weapon = Weapon(3, {"AP": 1})
dual_hand_weapon = Weapon(2)
fire_rifle = Weapon(2, {"Range": 18, "Rending": 0})
rifle = Weapon(1, {"Range": 18, "AP": 1})
crossbow = Weapon(1, {"Range": 24, "Rending": 0})

In [6]:
claw = Weapon(10)
stomp = Weapon(4, {"AP": 1})
giant_beast = Model(4, 3, [claw, stomp], rules={"Fear":2, "Regeneration":True, "Tough":12})
giant_beast_u = Unit([giant_beast], 1, name="Giant Beast")

In [7]:
import tqdm
def mc(func, times=10000):
    results = []
    for i in tqdm.tqdm(range(times)):
        results.append(func())
    return sum(results) / len(results)

In [8]:
def to_the_death(a, b):
    #pick a starter
    # A attacks b, b attacks a
    # until all dead
    
    first = a
    second = b
    
    # TODO add a shooting phase, start them 24" away from each other
    
    while True:
        # reset fatigued
        if "Fatigued" in a.rules:
            del a.rules["Fatigued"]
        if "Fatigued" in b.rules:
            del b.rules["Fatigued"]
        
        logger.debug("First melee")
        if "Wavering" not in first.rules:
            do_melee(first, second)
        else:
            del first.rules["Wavering"]
        if not a.alive:
            return 0
        if not b.alive:
            return 1
        
        logger.debug("Second melee")
        if "Wavering" not in second.rules:
            do_melee(second, first)
        else:
            del second.rules["Wavering"]
        if not a.alive:
            return 0
        if not b.alive:
            return 1

In [22]:
hand_weapon = Weapon(1)
halberd = Weapon(1, {"Rending": 0})
spear = Weapon(1, {"Counter": True})
dwarf_warrior = Model(4, 5, [hand_weapon], rules={"Shield Wall":0, "Slow":0})
warriors = Unit(models_for(dwarf_warrior, 20), 5, name="Warriors")
dwarf_halb = Model(4, 5, [halberd], rules={"Shield Wall":0, "Slow":0})
halb_warrs = Unit(models_for(dwarf_halb, 20), 5, name="Warriors")
dwarf_spear = Model(4, 5, [spear], rules={"Shield Wall":0, "Slow":0})
spear_warrs = Unit(models_for(dwarf_spear, 20), 5, name="Warriors")

In [32]:
heavy_hand_weapon = Weapon(1, {"AP": 1})
veteran = Model(3, 4, [heavy_hand_weapon], rules={"Fearless":0, "Slow":0})
veterans = Unit(models_for(veteran, 10), 10, name="Veterans")
heavy_spear = Weapon(1, {"AP": 1, "Counter": True})
spear_vet = Model(3, 4, [heavy_spear], rules={"Fearless":0, "Slow":0})
spear_vets = Unit(models_for(spear_vet, 10), 10, name="Spear Veterans")
veterans_3 = Unit(models_for(veteran, 15), 10, name="Veterans")


In [11]:
berserker = Model(4, 5, [dual_hand_weapon], rules={"Fearless":True, "Furious":True, "Slayer":True, "Slow":True})
berserkers = Unit(models_for(berserker, 10), 5, name="Berserkers")

In [12]:
rune_weapon = Weapon(1, {"Rending": 0})
iron_warrior = Model(3, 3, [rune_weapon], rules={"Shield Wall":True, "Slow":True})
iron_warriors = Unit(models_for(iron_warrior, 10), 5, name="Iron Warriors")

In [13]:
great_weapon = Weapon(2, {"AP": 2})
hammerer = Model(3, 4, [great_weapon], rules={"Slow":0})
hammerers = Unit(models_for(hammerer, 10), 5, name="Hammerers")

In [14]:
great_weapons = Weapon(6, {"AP": 2})
stomp = Weapon(4, {"AP": 1})
giant_construct = Model(3, 2, [great_weapons, stomp], rules={"Fear":2, "Slow":True, "Tough":12})
giant_construct_u = Unit([giant_construct], 1, name="Giant Construct")

In [15]:
stone_fists = Weapon(3, {"AP": 1})
golem = Model(4, 5, [stone_fists], rules={"Tough":3, "Slow":0})
golems = Unit(models_for(golem, 3), 3, name="Golems")
golems_c = Unit(models_for(golem, 6), 3, name="Golems")

In [16]:
# chivalrous kingdoms
throw_stones = Weapon(1, {"Range": 12})
peasant = Model(6, 6, [hand_weapon, throw_stones], rules={})
peasants = Unit(models_for(peasant, 20), 5, name="Peasants")

man_at_arms = Model(5, 5, [hand_weapon], rules={})
men_at_arms = Unit(models_for(man_at_arms, 20), 5, name="Men-At-Arms")

foot_knight = Model(4, 4, [heavy_hand_weapon], rules={})
foot_knights = Unit(models_for(foot_knight, 10), 5, name="Foot Knights")

lance = Weapon(1, {"Lance": True})
realm_knight = Model(4, 4, [lance], rules={"Fast": True, "Fearless": True, "Furious": True, "Impact": 1, "Joust": True})
realm_knights = Unit(models_for(realm_knight, 5), 5, name="Realm Knights")
realm_knights_c = Unit(models_for(realm_knight, 10), 5, name="Realm Knights")

In [39]:
giant_club = Weapon(6, {"AP": 2})
stomp = Weapon(4, {"AP": 1})
giant = Model(4, 3, [giant_club, stomp], rules={"Fear":2, "Slow":True, "Tough":12})
giant_u = Unit([giant], 1, name="Giant")

In [54]:
giant_beast_u, giant_construct_u, giant_u
a = halb_warrs
b = giant_u
with loggingcm(): to_the_death(copy.deepcopy(b), copy.deepcopy(a))

DEBUG:root:First melee
DEBUG:opr_logic:Unit(Giant (1/1)(0.0,0.0) strikes Unit(Warriors (20/20)(0.0,0.0)
DEBUG:opr_logic:10 attacks
DEBUG:opr_logic:6 hits
DEBUG:opr_logic:3 wounds
DEBUG:opr_logic:Unit(Warriors (17/20)(0.0,0.0) strikes Unit(Giant f(1/1)(0.0,0.0)
DEBUG:opr_logic:10 attacks
DEBUG:opr_logic:5 hits
DEBUG:opr_logic:4 wounds
DEBUG:opr_logic:scores: a: 4, d: 7
DEBUG:opr_structs:Giant doing morale test
DEBUG:root:Second melee
DEBUG:opr_logic:Unit(Warriors f(17/20)(0.0,0.0) strikes Unit(Giant f(1/1 w4)(0.0,0.0)
DEBUG:opr_logic:10 attacks
DEBUG:opr_logic:1 hits
DEBUG:opr_logic:1 wounds
DEBUG:opr_logic:Unit(Giant f(1/1 w5)(0.0,0.0) strikes Unit(Warriors f(17/20)(0.0,0.0)
DEBUG:opr_logic:10 attacks
DEBUG:opr_logic:1 hits
DEBUG:opr_logic:1 wounds
DEBUG:opr_logic:scores: a: 4, d: 2
DEBUG:opr_structs:Giant doing morale test
DEBUG:opr_structs:Giant failed morale test
DEBUG:root:First melee
DEBUG:root:Second melee
DEBUG:opr_logic:Unit(Warriors (16/20)(0.0,0.0) strikes Unit(Giant (1/1 w5)

In [55]:
a_first = mc(lambda: to_the_death(copy.deepcopy(a), copy.deepcopy(b)), times=10000)
b_first = 1 - mc(lambda: to_the_death(copy.deepcopy(b), copy.deepcopy(a)), times=10000)
(a_first + b_first) / 2, a_first, b_first

100%|█████████████████████████████████████████████████████████| 10000/10000 [00:06<00:00, 1457.89it/s]


(0.49119999999999997, 0.5519, 0.4305)

In [ ]:
# def mc_many

Without slayer: 0.018
With slayer: 0.17
berserkers: (0.3737, 0.15859999999999996, 0.26615)
with fearless: (0.3871, 0.16779999999999995, 0.27745)
Iron Warriors: 0.367
Iron Warriors with sw: 0.754 - both first = 0.50 (0.7476, 0.688, 0.7178)
Hammerers: 0.503 - both: (0.505, 0.2532, 0.3791)

vs giant construct: 
warriors vs giant construct: (0.003699999999999984, 0.0061, 0.0012999999999999678)
rending warriors: (0.14500000000000002, 0.1772, 0.11280000000000001)
counter warriors: (0.0030500000000000227, 0.0045, 0.0016000000000000458)
veterans vs giant construct: (0.0013999999999999724, 0.0023, 0.0004999999999999449)
Berserkers vs giant construct: (0.22865, 0.3642, 0.09309999999999996)
Hammerers vs giant construct: (0.2149, 0.3515, 0.07830000000000004)
Iron warriors vs giant construct: (0.0736, 0.0979, 0.04930000000000001)
spear_vets: (0.0020499999999999954, 0.002, 0.0020999999999999908)
veterans_3: (0.03749999999999998, 0.0584, 0.016599999999999948)

vs warriors:
warriors: (0.5018499999999999, 0.5678, 0.43589999999999995)
halb_warrs: (0.7605, 0.8168, 0.7041999999999999)
spear_warrs: (0.5393, 0.5854, 0.49319999999999997)
Veterans: (0.12745, 0.1848, 0.07010000000000005)
berserkers: (0.35650000000000004, 0.4711, 0.2419)
iron_warriors: (0.84885, 0.8907, 0.8069999999999999)
hammerers: (0.8624499999999999, 0.9642, 0.7606999999999999)
golems: (0.007200000000000018, 0.0107, 0.0037000000000000366)
golems_c: (0.6537999999999999, 0.7632, 0.5444)
peasants: (0.0003, 0.0006, 0.0)
men_at_arms: (0.05234999999999998, 0.0711, 0.03359999999999996)
foot_knights: (0.04919999999999997, 0.0717, 0.026699999999999946)
realm_knights: (0.02479999999999999, 0.0383, 0.011299999999999977)
realm_knights_c: (0.6936, 0.8551, 0.5321)
spear_vets: (0.19534999999999997, 0.2129, 0.17779999999999996)
veterans_3: (0.78435, 0.8921, 0.6766)

vs veterans:
men_at_arms: (0.40115, 0.5128, 0.2895)
vs v3: 

vs Giant:
warriors: (0.15254999999999996, 0.1897, 0.11539999999999995)
veterans: (0.03445, 0.0548, 0.014100000000000001)
spear_vets: (0.06, 0.0605, 0.0595)
berserkers: (0.6508, 0.8216, 0.48)
iron_warriors: (0.33155, 0.395, 0.2681)
hammerers: (0.6374500000000001, 0.8328, 0.44210000000000005)
halb_warrs: (0.49119999999999997, 0.5519, 0.4305)